File: src/main/java/rbq/wtf/lycoris/client/transformer/TransformManager.java
package rbq.wtf.lycoris.client.transformer;

import rbq.wtf.lycoris.agent.LycorisAgent;
import rbq.wtf.lycoris.agent.instrument.impl.InstrumentationImpl;
import rbq.wtf.lycoris.client.Client;
import rbq.wtf.lycoris.client.transformer.transformers.*;
import rbq.wtf.lycoris.client.utils.FileUtils;
import rbq.wtf.lycoris.client.utils.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class TransformManager {
    public static List<ClassTransformer> transformers = new ArrayList<>();
    public static List<BackTransformer> backTransformers = new ArrayList<>();

    public static boolean transformed = false;

    public static void init() {
        if (transformed)
            throw new RuntimeException("Classes has been transformed");
        Logger.info("Start Initialize Transforms", "Transformer");
        transformers.add(new MinecraftTransformer());
        transformers.add(new GuiIngameTransformer());
        transformers.add(new KeyBindingTransformer());
        transformers.add(new EntityPlayerSPTransformer());
        transformers.add(new EntityRendererTransformer());
        transformers.add(new NetworkManagerTransformer());
        doTransform();
        Logger.info("Transforms Initialized Successful", "Transformer");
    }

    public static void doTransform() {
        //Load Native to use JVMTI to call onTransform
        Logger.info("Start Initialize Native", "Native");
        Logger.info("Load Native: " + Client.JVMTILib.getFile().getName(), "Native");
        System.load(Client.JVMTILib.getFile().toString());
        Logger.info("Native Initialized Successful", "Native");
        //re-transform Class
        if (!transformed) {
            for (ClassTransformer classTransformer : transformers) {
                Logger.info("Start Transformer " + classTransformer.getTargetClass().getCanonicalName(), "Transformer");
                int error = LycorisAgent.retransformclass(new InstrumentationImpl(), classTransformer.getTargetClass());
                Logger.info("Transformed Class " + classTransformer.getTargetClass().getCanonicalName() + " " + JVMTIError.parse(error), "Transformer");
            }
            transformed = true;
        } else {
            for (ClassTransformer classTransformer : backTransformers) {
                Logger.info("Start Transformer " + classTransformer.getTargetClass().getCanonicalName(), "TransformBack");
                int error = LycorisAgent.retransformclass(new InstrumentationImpl(), classTransformer.getTargetClass());
                Logger.info("Transformed Class " + classTransformer.getTargetClass().getCanonicalName() + " " + JVMTIError.parse(error), "TransformBack");
            }
            backTransformers.clear();
            transformed = false;
        }
    }

    public static byte[] onTransform(Class<?> clazz, byte[] original_class_bytes) {
        if (clazz == null || clazz.getCanonicalName().equals("null")) {
            Logger.warning("NULL Class", "Transformer");
            return original_class_bytes;
        }
        if (original_class_bytes == null) {
            Logger.warning("NULL Class Bytes", "Transformer");
            return null;
        }
        byte[] class_bytes = null;
        if (!transformed) {
            for (ClassTransformer transformer : transformers) {
                if (transformer.getTargetClass().equals(clazz)) {
                    class_bytes = transformer.transform(original_class_bytes);
                    backTransformers.add(new BackTransformer(clazz, original_class_bytes));
                    Logger.debug("Transform" + clazz.getCanonicalName() + " Successful", "Transformer");
                }
            }
        } else {
            for (ClassTransformer transformer : backTransformers) {
                if (transformer.getTargetClass().equals(clazz)) {
                    class_bytes = transformer.transform(original_class_bytes);
                    Logger.debug("Transform Back" + clazz.getCanonicalName() + " Successful", "Transformer");
                }
            }
        }
        if (class_bytes == null) {
            Logger.error("Transform " + clazz.getCanonicalName() + " return null", "Transformer");
            return original_class_bytes;
        }
        if (Client.developEnv) {
            try {
                FileUtils.writeFile(Client.runPath.getParent().resolve("debug/" + clazz.getCanonicalName().replaceAll("/", ".") + (transformed ? "_back" : "") + ".class").toFile(), class_bytes);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return class_bytes;
    }
}

File: src/main/java/rbq/wtf/lycoris/client/transformer/TransformManager.kt
package rbq.wtf.lycoris.client.transformer

import rbq.wtf.lycoris.agent.LycorisAgent
import rbq.wtf.lycoris.agent.instrument.impl.InstrumentationImpl
import rbq.wtf.lycoris.client.Client
import rbq.wtf.lycoris.client.transformer.transformers.*
import rbq.wtf.lycoris.client.utils.FileUtils.writeFile
import rbq.wtf.lycoris.client.utils.Logger
import java.io.IOException

object TransformManager {
    var transformers: MutableList<ClassTransformer> = ArrayList()
    var backTransformers: MutableList<BackTransformer> = ArrayList()
    var transformed = false

    fun init() {
        if (transformed) throw RuntimeException("Classes has been transformed")
        Logger.info("Start Initialize Transforms", "Transformer")
        transformers.add(MinecraftTransformer())
        transformers.add(GuiIngameTransformer())
        transformers.add(KeyBindingTransformer())
        transformers.add(EntityPlayerSPTransformer())
        transformers.add(EntityRendererTransformer())
        transformers.add(NetworkManagerTransformer())
        doTransform()
        Logger.info("Transforms Initialized Successful", "Transformer")
    }

    fun doTransform() {
        //Load Native to use JVMTI to call onTransform
        Logger.info("Start Initialize Native", "Native")
        Logger.info("Load Native: " + Client.JVMTILib.file.name, "Native")
        System.load(Client.JVMTILib.file.toString())
        Logger.info("Native Initialized Successful", "Native")
        //re-transform Class
        if (!transformed) {
            for (classTransformer in transformers) {
                Logger.info("Start Transformer " + classTransformer.targetClass.canonicalName, "Transformer")
                val error = LycorisAgent.retransformclass(InstrumentationImpl(), classTransformer.targetClass)
                Logger.info(
                    "Transformed Class " + classTransformer.targetClass.canonicalName + " " + JVMTIError.parse(error),
                    "Transformer"
                )
            }
            transformed = true
        } else {
            for (classTransformer in backTransformers) {
                Logger.info("Start Transformer " + classTransformer.targetClass.canonicalName, "TransformBack")
                val error = LycorisAgent.retransformclass(InstrumentationImpl(), classTransformer.targetClass)
                Logger.info(
                    "Transformed Class " + classTransformer.targetClass.canonicalName + " " + JVMTIError.parse(error),
                    "TransformBack"
                )
            }
            backTransformers.clear()
            transformed = false
        }
    }

    @JvmStatic
    fun onTransform(clazz: Class<*>?, original_class_bytes: ByteArray?): ByteArray? {
        if (clazz == null || clazz.canonicalName == "null") {
            Logger.warning("NULL Class", "Transformer")
            return original_class_bytes
        }
        if (original_class_bytes == null) {
            Logger.warning("NULL Class Bytes", "Transformer")
            return null
        }
        var classBytes = ByteArray(0)
        if (!transformed) {
            for (transformer in transformers) {
                if (transformer.targetClass == clazz) {
                    classBytes = transformer.transform(original_class_bytes)
                    backTransformers.add(BackTransformer(clazz, original_class_bytes))
                    Logger.debug("Transform" + clazz.canonicalName + " Successful", "Transformer")
                }
            }
        } else {
            for (transformer in backTransformers) {
                if (transformer.targetClass == clazz) {
                    classBytes = transformer.transform(original_class_bytes)
                    Logger.debug("Transform" + clazz.canonicalName + " Successful", "TransformBack")
                }
            }
        }
//        if (classBytes == null) {
//            Logger.error("Transform " + clazz.canonicalName + " return null", "Transformer")
//            return original_class_bytes
//        }
        if (Client.developEnv) {
            try {
                writeFile(Client.runPath.parent.resolve("debug/" + clazz.canonicalName.replace("/", ".") + (if (transformed) "_back" else "") + ".class").toFile(), classBytes)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        return classBytes
    }
}